<!-- ===============================
Couverture
=============================== -->
<div style="background-color:#004080; color:white; padding:20px; border-radius:10px; text-align:center;">
  <h1>Audit linguistique du sous-ensemble Wolof</h1>
  <h3>Dataset MURI-IT — Reverse Instruction</h3>
  <p style="font-size:16px;">Auteur : Marème DIOP | Date : 21/02/2026</p>
  <p style="font-size:14px; font-style:italic;">
  Ce rapport présente une évaluation méthodologique du sous-ensemble étiqueté « Wolof »
  dans le dataset MURI-IT, afin d’examiner la cohérence entre l’étiquetage linguistique
  et le contenu effectif des données.
  </p>
</div>

---

<!-- ===============================
Contexte scientifique
=============================== -->
<h2 id="contexte" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Contexte scientifique
</h2>

<p>
Le dataset <strong>MURI-IT</strong> repose sur un mécanisme de génération d’instructions par
<em>Reverse Instruction</em>, appliqué à des corpus multilingues. Dans ce type de pipeline,
la validité linguistique des sous-ensembles générés dépend étroitement du système
de détection automatique de langue (Language Identification — LID) utilisé pour
l’étiquetage et la validation des données.
</p>

<p>
Dans le cas des langues à faibles ressources, telles que le wolof,
les performances des détecteurs généralistes peuvent varier,
et une validation insuffisamment robuste peut entraîner une discordance
entre l’étiquetage linguistique et le contenu réel des textes.
Une telle discordance affecte directement la qualité du corpus
et la fiabilité des étapes ultérieures d’entraînement ou d’évaluation.
</p>

<p>
Le présent travail s’inscrit dans une démarche d’audit méthodologique
visant à examiner la cohérence linguistique du sous-ensemble
étiqueté « Wolof » dans MURI-IT.
L’analyse porte conjointement sur les instructions générées
et sur les documents sources associés.
</p>

---

<!-- ===============================
Objectif et hypothèses
=============================== -->
<h2 id="objectif" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;"> Objectif et hypothèses
</h2>

<p>
L’objectif de cette étude est d’évaluer empiriquement
si le sous-ensemble étiqueté « Wolof »
correspond effectivement à un corpus majoritairement rédigé en wolof,
et d’examiner la robustesse de son étiquetage linguistique.
</p>

<p>
Afin de structurer l’analyse, les hypothèses suivantes sont retenues :
</p>

<ul>
<li>
<strong>H1 — Discordance linguistique:
  </strong>
Le sous-ensemble étiqueté « Wolof » présente une discordance significative
entre l’étiquetage linguistique et le contenu réel des textes.
</li>

<li>
<strong>H2 — Cohérence de validation :</strong>
L’étiquetage linguistique du sous-ensemble est cohérent
avec les prédictions de détecteurs de langue indépendants et robustes.
</li>

<li>
<strong>H3 — Détectabilité du signal :</strong>
Si le sous-ensemble est effectivement majoritairement rédigé en wolof,
alors un signal linguistique wolof devrait être détectable de manière robuste
par des classifieurs multilingues indépendants,
avec des niveaux de confiance substantiels.
</li>
</ul>

<p>
L’analyse qui suit mobilise plusieurs détecteurs de langue indépendants
afin de tester ces hypothèses de manière quantitative et comparative.</p>

<!-- ===============================
Cadre méthodologique
=============================== -->
<h2 id="methodologie" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Cadre méthodologique
</h2>

<p>
L’évaluation repose sur une stratégie hiérarchique de détection linguistique
visant à tester la robustesse du signal wolof dans un espace linguistique ouvert,
puis dans un espace spécialisé.
</p>

<p>
L’utilisation exclusive d’un détecteur spécialisé pour les langues africaines
n’est pas suffisante dans ce contexte. En effet, ces modèles opèrent dans un
espace restreint et ne couvrent pas l’ensemble des langues globalement dominantes,
notamment l’anglais et le français. Lorsqu’un texte appartient à une langue
non couverte par le modèle, celui-ci peut produire une attribution contrainte
vers la classe la moins improbable, sans que cela ne reflète un signal linguistique réel.
</p>

<p>
Afin d’éviter ce biais structurel, l’analyse procède en trois étapes :
</p>

<ul>
<li>
<strong>Étape 1 — Filtrage multilingue global :</strong>
Application d’un classifieur couvrant un large ensemble de langues,
afin d’identifier les langues dominantes dans un espace ouvert
et de détecter la présence éventuelle de langues non africaines.
</li>

<li>
<strong>Étape 2 — Calibration par seuil de confiance :</strong>
Séparation des prédictions robustes des cas ambigus,
sur la base des niveaux de confiance fournis par le classifieur global.
</li>

<li>
<strong>Étape 3 — Analyse spécialisée africaine :</strong>
Application d’un détecteur spécialisé sur les cas résiduels,
afin d’examiner la présence d’un signal wolof dans un espace linguistique restreint.
</li>
</ul>

<p>
Cette approche hiérarchique permet de distinguer les langues
clairement identifiables à l’échelle globale des cas nécessitant
une analyse intra-africaine plus fine,
et d’évaluer la détectabilité effective du signal wolof
dans le sous-ensemble étudié.
</p>

<!-- ===============================
Table des matières
=============================== -->
<div style="background-color:#f0f0f0; padding:15px; border-radius:8px; line-height:1.6;">
<h2 style="color:#004080;">📑 Table des matières</h2>

<ol>
<li><a href="#contexte">Contexte scientifique</a></li>

<li><a href="#objectif">Objectif et hypothèses</a></li>

<li><a href="#dataset">Présentation du sous-ensemble étudié</a></li>
<li><a href="#models">Modèles de détection de langue utilisés</a></li>

<li><a href="#fasttext_detection">Détection multilingue globale (fastText)</a></li>

<li><a href="#afrolid_analysis">Analyse spécialisée des cas ambigus (AfroLID)</a></li>

<li><a href="#triangulation">Analyse comparative et triangulation</a></li>

<li><a href="#discussion">Discussion méthodologique</a></li>

<li><a href="#conclusion">Conclusion générale</a></li>
</ol>

</div>

<!-- ===============================
Présentation du sous-ensemble étudié
=============================== -->
<h2 id="dataset" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Présentation du sous-ensemble étudié
</h2>

<p>
Le présent audit porte sur le sous-ensemble étiqueté <strong>« Wolof »</strong>
au sein du dataset <strong>MURI-IT</strong>, issu d’un pipeline de génération
par <em>Reverse Instruction</em>. Dans ce mécanisme, un document source est sélectionné,
traduit vers l’anglais, associé à une instruction générée automatiquement,
puis retraduit dans la langue cible.
</p>

<p>
Chaque exemple du sous-ensemble comprend deux composantes distinctes :
</p>

<ul>
<li><strong>Instruction (input)</strong> — requête générée automatiquement.</li>
<li><strong>Output (document source)</strong> — texte issu du corpus initial.</li>
</ul>

<p>
La distinction entre ces deux composantes est maintenue dans l’analyse,
afin d’examiner séparément les effets potentiels du mécanisme de génération
sur la validité linguistique.
</p>

---

<h3 style="color:#004080;">Taille et structure du sous-ensemble</h3>

<ul>
<li><strong>Nombre total d’exemples analysés :</strong> 4457</li>
<li><strong>Nombre d’instructions :</strong> 4457</li>
<li><strong>Nombre d’outputs :</strong> 4457</li>
</ul>

<p>
La correspondance un-à-un entre instructions et outputs est conservée
dans l’ensemble analysé.
</p>

---

<h3 style="color:#004080;">Caractéristiques textuelles globales</h3>

<p>
Les instructions présentent une longueur moyenne d’environ
<strong>144 caractères</strong> (≈ 25 mots), correspondant à des requêtes relativement brèves.
Les outputs présentent une longueur moyenne d’environ
<strong>481 caractères</strong>, avec une distribution plus hétérogène.
</p>

<p>
Ces caractéristiques descriptives fournissent un cadre contextuel
à l’analyse linguistique présentée dans les sections suivantes.
</p>

In [ ]:
# ============================================================
# Chargement des bibliothèques
# ============================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datasets import load_dataset

pd.set_option("display.max_colwidth", 200)

print("Bibliothèques chargées avec succès.")

Bibliothèques chargées avec succès.


In [ ]:
# ============================================================
# Chargement du dataset MURI-IT
# ============================================================

# Option 1 : Chargement depuis HuggingFace

df_wolof = pd.read_csv("/content/muri_wolof.csv")

print("Dataset chargé.")


Dataset chargé.


In [ ]:


print(f"Nombre d'exemples étiquetés Wolof : {len(df_wolof)}")

Nombre d'exemples étiquetés Wolof : 4457


In [ ]:
# ============================================================
# Séparation des instructions et des outputs
# ============================================================

instructions = df_wolof["input"].astype(str)
outputs = df_wolof["output"].astype(str)

print("Séparation effectuée.")
print(f"Nombre d'instructions : {len(instructions)}")
print(f"Nombre d'outputs : {len(outputs)}")

Séparation effectuée.
Nombre d'instructions : 4457
Nombre d'outputs : 4457


In [ ]:
# ============================================================
# Statistiques descriptives préliminaires
# ============================================================

def compute_basic_stats(series, name):
    lengths_chars = series.apply(len)
    lengths_words = series.apply(lambda x: len(x.split()))

    print(f"\nStatistiques pour : {name}")
    print("-" * 40)
    print(f"Longueur moyenne (caractères) : {lengths_chars.mean():.2f}")
    print(f"Longueur médiane (caractères) : {lengths_chars.median():.2f}")
    print(f"Longueur moyenne (mots) : {lengths_words.mean():.2f}")
    print(f"Longueur médiane (mots) : {lengths_words.median():.2f}")

compute_basic_stats(instructions, "Instructions")
compute_basic_stats(outputs, "Outputs")


Statistiques pour : Instructions
----------------------------------------
Longueur moyenne (caractères) : 144.06
Longueur médiane (caractères) : 136.00
Longueur moyenne (mots) : 25.30
Longueur médiane (mots) : 24.00

Statistiques pour : Outputs
----------------------------------------
Longueur moyenne (caractères) : 481.49
Longueur médiane (caractères) : 105.00
Longueur moyenne (mots) : 92.38
Longueur médiane (mots) : 21.00


<!-- ===============================
Modèles de détection utilisés
=============================== -->
<h2 id="models" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Modèles de détection de langue utilisés
</h2>

<p>
L’analyse repose sur une stratégie hiérarchique mobilisant deux détecteurs complémentaires :
</p>

<ul>
<li><strong>fastText LID (176 langues)</strong> — utilisé en première étape pour une détection multilingue globale, incluant les langues à forte ressource (anglais, français, etc.).</li>
<li><strong>AfroLID</strong> — utilisé en seconde étape pour une analyse spécialisée des langues africaines, notamment le wolof.</li>
</ul>

<p>
fastText permet d’identifier les langues dominantes non couvertes par AfroLID,
tandis qu’AfroLID est mobilisé pour affiner l’analyse sur les cas résiduels.
Cette combinaison vise à limiter les biais liés à la couverture linguistique des modèles.
</p>

<p>
Le choix d’AfroLID s’appuie sur une évaluation comparative préalable des détecteurs de langue pour le wolof :
<br>
<a href="https://github.com/MaDi-12/wolof-language-detection-eval" target="_blank">
Benchmark des détecteurs de langue pour le Wolof
</a>
</p>

<p>
La section suivante présente les résultats de la détection multilingue globale.
</p>

<!-- ===============================
Détection multilingue globale
=============================== -->
<h2 id="fasttext_detection" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Détection multilingue globale
</h2>

<p>
Une première analyse est réalisée à l’aide du classifieur multilingue
<strong>fastText LID (176 langues)</strong>, afin d’identifier
la distribution globale des langues présentes dans le sous-ensemble
étiqueté « Wolof ».
</p>

<p>
Cette étape vise à tester l’existence d’un signal linguistique
dominant compatible avec l’étiquetage initial,
et à détecter d’éventuelles langues à forte ressource
susceptibles de contaminer le sous-ensemble.
</p>

---

<h3 style="color:#004080;">Application d’un seuil de confiance</h3>

<p>
Afin de distinguer les prédictions robustes des cas incertains,
un seuil de confiance <strong>τ = 0.7</strong> est appliqué
aux probabilités retournées par fastText.
</p>

<p>
Les instances dont la probabilité maximale est supérieure ou égale à τ
sont considérées comme des prédictions à haute confiance.
Les autres sont classées comme cas ambigus
et feront l’objet d’une analyse complémentaire.
</p>

---

<h3 style="color:#004080;">Résultats globaux</h3>

<p>
La distribution des langues détectées par fastText
est présentée ci-dessous.
</p>

In [ ]:
!pip install transformers==4.37.2 --quiet

In [ ]:
# ===============================
# Chargement du détecteur global (fastText LID 176)
# ===============================

!pip install numpy==1.26.4 --quiet
!pip install fasttext --quiet
!wget -q https://dl.fbaipublicfiles.com/fasttext/supervised-models/lid.176.bin



In [ ]:
import fasttext

ft_model = fasttext.load_model("lid.176.bin")


In [ ]:
# ===============================
# Détection de langue avec fastText
# ===============================

from tqdm import tqdm
import pandas as pd

def detect_language_fasttext(text):
    try:
        labels, probabilities = ft_model.predict(str(text).replace("\n", " "), k=1)

        language = labels[0].replace("__label__", "")
        confidence = float(probabilities[0])

        return language, confidence

    except Exception as e:
        return "error", 0.0

In [ ]:
from tqdm import tqdm
import pandas as pd

tqdm.pandas()

df_wolof[["ft_pred_lang", "ft_confidence"]] = (
    df_wolof["input"]
    .progress_apply(lambda x: pd.Series(detect_language_fasttext(x)))
)

100%|██████████| 4457/4457 [00:00<00:00, 7371.11it/s] 


In [ ]:
# ===============================
# Résumé statistique — fastText (Instructions)
# ===============================

total = len(df_wolof)

# Distribution complète des langues
language_counts = df_wolof["ft_pred_lang"].value_counts()
language_percentages = df_wolof["ft_pred_lang"].value_counts(normalize=True) * 100

summary_df = pd.DataFrame({
    "Nombre": language_counts,
    "Pourcentage (%)": language_percentages.round(2)
})

print("📊 Distribution complète des langues détectées (fastText)")
display(summary_df.head(10))

# Taux de wolof
wolof_detected = language_counts.get("wo", 0)
wolof_rate = wolof_detected / total

print("\n" + "-"*50)
print("📌 Indicateurs clés")
print(f"Nombre total d’instructions analysées : {total}")
print(f"Nombre détecté comme wolof : {wolof_detected}")
print(f"Taux de wolof détecté : {wolof_rate:.2%}")
print(f"Taux estimé de non-wolof : {(1 - wolof_rate):.2%}")

# Analyse des scores de confiance
print("\n" + "-"*50)
print("📈 Statistiques des scores de confiance")
display(df_wolof["ft_confidence"].describe())

📊 Distribution complète des langues détectées (fastText)


,Nombre,Pourcentage (%)
ft_pred_lang,,
en,1924,43.17
fr,444,9.96
cy,140,3.14
it,107,2.40
tl,104,2.33
zh,93,2.09
sw,84,1.88
ur,84,1.88
es,74,1.66



--------------------------------------------------
📌 Indicateurs clés
Nombre total d’instructions analysées : 4457
Nombre détecté comme wolof : 0
Taux de wolof détecté : 0.00%
Taux estimé de non-wolof : 100.00%

--------------------------------------------------
📈 Statistiques des scores de confiance


,ft_confidence
count,4457.000000
mean,0.573474
std,0.333502
min,0.045478
25%,0.246158
50%,0.528536
75%,0.934624
max,1.000067


In [ ]:
# ===============================
# Séparation par seuil de confiance (fastText)
# ===============================

CONFIDENCE_THRESHOLD = 0.7

# Marquage haute / basse confiance
df_wolof["ft_high_confidence"] = df_wolof["ft_confidence"] >= CONFIDENCE_THRESHOLD

# Séparation des sous-ensembles
df_high_conf = df_wolof[df_wolof["ft_high_confidence"]].copy()
df_low_conf = df_wolof[~df_wolof["ft_high_confidence"]].copy()

# Statistiques
total = len(df_wolof)
high_count = len(df_high_conf)
low_count = len(df_low_conf)

print("📊 Résumé de la séparation par seuil")
print("-" * 50)
print(f"Seuil de confiance τ = {CONFIDENCE_THRESHOLD}")
print(f"Total instructions : {total}")
print(f"Haute confiance (≥ τ) : {high_count} ({high_count/total:.2%})")
print(f"Cas ambigus (< τ) : {low_count} ({low_count/total:.2%})")

📊 Résumé de la séparation par seuil
--------------------------------------------------
Seuil de confiance τ = 0.7
Total instructions : 4457
Haute confiance (≥ τ) : 1912 (42.90%)
Cas ambigus (< τ) : 2545 (57.10%)


In [ ]:
df_low_conf.sample(20)

,input,output,dataset_name,subdataset_name,language,split,language_name,ft_pred_lang,ft_confidence,ft_high_confidence,afrolid_lang,afrolid_conf
719,"Lu tax lu baax ci kaw guumtu Senegal yi, te xa...","Ceeb u jën di lekk gu siiw ca Senegaal, ñu ëpp...",MRI,wikipedia,wol,train,Wolof,jbo,0.138281,False,[wol],[0.0019269605170805942]
1520,"Lu tax lu ngi leen diw ci ""Ndiané"" ? Li ko gën...",Ndiané : Pentanemus quinquarius (dundat yi)\n\...,MRI,wikipedia,wol,train,Wolof,cy,0.082712,False,[wol],[0.0019266005878743058]
1417,Wọ́n di akọ́ṣẹ́mọṣẹ́ ògbóntarìgì nínú kí a gé ...,Danoo aay ci dagg cër ngir muccal ñi feebar ci...,xP3,NaN,wol,train,Wolof,yo,0.500671,False,[yor],[0.0019304487696025362]
4383,Vahlaseri va le Mumbai va fike hi xikepe hi No...,Ci def congu gi agsi nanu ci gaal ci ñaar fukk...,xP3,NaN,wol,train,Wolof,en,0.163509,False,[tso],[0.0009652238952754016]
3199,Text in French: Par la mort (drame en deux act...,Yoonu murit (33 ciy xët ci biir),xP3,NaN,wol,train,Wolof,en,0.222521,False,[wol],[0.001232694014619216]
4013,Teen leen ci taariki Senegal. Taariix bi ko gë...,"Lat Joor Joob mi ngi juddoo ca Kër Amadu Yala,...",MRI,wikipedia,wol,train,Wolof,cy,0.122156,False,[wol],[0.001926679178554579]
3851,"Defar ci réewum Burkina Faso. Ci lu baax la, d...","Mi ngi nekk ci goxub Afrig, ci diwaanu sowwu A...",MRI,wikipedia,wol,train,Wolof,cy,0.101937,False,[wol],[0.0019273060629266569]
78,A text in Tswana: Popelelo e simolotse ka 11:0...,Fippu gi door ci 11:00 ci Whitehall ci diggant...,xP3,NaN,wol,train,Wolof,sw,0.345993,False,[tsn],[0.001884659399857392]
897,"A text in Akan: Cherry blossom ho nsusueɛ, a w...","Xool tóor-tóoru cerisier, ñu xamee ko ci hanam...",xP3,NaN,wol,train,Wolof,en,0.520206,False,[twi],[0.0019300611092777325]
109,"Gantsi, o tshwanelwa ke go reka nomore ya lefa...","Yénn say, dangay wara jeend lu beeri lu wuté l...",xP3,NaN,wol,train,Wolof,en,0.465163,False,[tsn],[0.0018796743573369207]


<!-- ===============================
Interprétation — Détection multilingue globale
=============================== -->
<h3 style="color:#004080; border-bottom:2px solid #004080; padding-bottom:4px;" id="">
Interprétation des résultats — Détection globale (fastText)
</h3>

<p>
L’analyse multilingue réalisée sur les 4 457 instructions
montre une distribution linguistique dominée par des langues
à forte ressource.
</p>

<ul>
<li>Anglais : 43,17 % (1 924 instances)</li>
<li>Français : 9,96 % (444 instances)</li>
<li>Autres langues diverses : proportions inférieures à 4 % chacune</li>
</ul>

<p>
Aucune instance n’est détectée comme wolof par fastText,
soit un taux de wolof détecté de 0,00 %.
</p>

---

<h4 style="color:#004080;">Analyse des niveaux de confiance</h4>

<p>
Le score moyen de confiance est de 0,57 (écart-type = 0,33),
avec une médiane à 0,53.
Le troisième quartile atteint 0,93,
indiquant qu’une fraction substantielle des prédictions
présente une probabilité élevée.
</p>

<p>
Après application d’un seuil τ = 0,7 :
</p>

<ul>
<li>42,90 % des instructions (1 912) sont classées à haute confiance</li>
<li>57,10 % (2 545) constituent des cas ambigus</li>
</ul>

---

<h4 style="color:#004080;">Interprétation méthodologique</h4>

<p>
La dominance de l’anglais et du français,
associée à l’absence totale de détection du wolof,
indique qu’aucun signal wolof robuste
n’est identifiable par un classifieur multilingue global.
</p>

<p>
Les prédictions à haute confiance suggèrent que
près de la moitié du sous-ensemble est linguistiquement
cohérente avec des langues non-wolof.
</p>

<p>
La proportion majoritaire de cas ambigus (57,10 %)
justifie une analyse complémentaire,
afin de déterminer si un signal wolof
faiblement capté par fastText
pourrait être détecté par un modèle spécialisé.
</p>

<p>
Ces résultats fournissent un premier élément empirique
en faveur de l’hypothèse H1
(discordance entre étiquetage et contenu),
tout en nécessitant une vérification
dans le cadre de l’hypothèse H3.
</p>

<!-- ===============================
Analyse spécialisée des cas ambigus
=============================== -->
<h2 id="afrolid_analysis" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Analyse spécialisée des cas ambigus
</h2>

<p>
Les résultats obtenus avec fastText montrent qu’une proportion majoritaire
des instructions (57,10 %) présente un niveau de confiance inférieur au seuil
τ = 0,7.
</p>

<p>
Ces 2 545 instances constituent un sous-ensemble
linguistiquement incertain,
pour lequel la détection multilingue globale
ne permet pas de conclure de manière robuste.
</p>

<p>
Afin d’examiner l’existence éventuelle
d’un signal wolof faiblement capté par fastText,
nous appliquons AfroLID
à ces cas ambigus uniquement.
</p>

<p>
Cette étape vise à tester si un détecteur spécialisé
dans les langues africaines
identifie une proportion substantielle d’exemples
comme relevant du wolof.
</p>

<p>
Les résultats sont présentés ci-dessous.
</p>

In [ ]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification


The cache for model files in Transformers v4.22.0 has been updated. Migrating your old cache. This is a one-time only operation. You can interrupt this and resume the migration later on by calling `transformers.utils.move_cache()`.


0it [00:00, ?it/s]

In [ ]:
from transformers import pipeline
from tqdm import tqdm
tqdm.pandas()

classifier_afrolid = pipeline(
    "text-classification",
    model="UBC-NLP/afrolid_1.5",
    return_all_scores=True
)

print("AfroLID 1.5 chargé depuis HuggingFace.")

/usr/local/lib/python3.12/dist-packages/huggingface_hub/file_download.py:949: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(


AfroLID 1.5 chargé depuis HuggingFace.


/usr/local/lib/python3.12/dist-packages/transformers/pipelines/text_classification.py:105: UserWarning: `return_all_scores` is now deprecated,  if want a similar functionality use `top_k=None` instead of `return_all_scores=True` or `top_k=1` instead of `return_all_scores=False`.
  warnings.warn(


In [ ]:
# ============================================================
# Fonction de prédiction robuste avec gestion des textes longs
# ============================================================

def split_text(text, max_chars=300):
    """
    Découpe un texte en segments de longueur max_chars.
    """
    return [text[i:i+max_chars] for i in range(0, len(text), max_chars)]


def predict_afrolid(texts, max_chars=300):
    predicted_languages = []
    confidence_scores = []

    for text in tqdm(texts):
        if len(text) <= max_chars:
            results = classifier_afrolid(text)[0]
        else:
            segments = split_text(text, max_chars=max_chars)
            segment_results = []

            for seg in segments:
                seg_pred = classifier_afrolid(seg)[0]
                segment_results.append(seg_pred)

            # Agrégation des segments
            results = []
            for label_scores in segment_results:
                results.extend(label_scores)

        # Agrégation des scores
        label_scores_dict = {}
        for entry in results:
            label = entry['label']
            score = entry['score']
            label_scores_dict[label] = label_scores_dict.get(label, 0) + score

        # Langue majoritaire pondérée
        best_label = max(label_scores_dict, key=label_scores_dict.get)
        best_score = label_scores_dict[best_label] / len(results)

        predicted_languages.append(best_label)
        confidence_scores.append(best_score)

    return predicted_languages, confidence_scores

In [ ]:
# ===============================
# Vérification AfroLID sur cas ambigus et analyse des résultats
# ===============================

afrolid_predicted_langs, afrolid_confidence_scores = predict_afrolid(df_low_conf["input"].tolist())

df_low_conf["afrolid_lang"] = afrolid_predicted_langs
df_low_conf["afrolid_conf"] = afrolid_confidence_scores

print("✅ Vérification AfroLID des cas ambigus terminée.")

# ===============================
# Analyse des résultats AfroLID pour les cas ambigus
# ===============================

total_low_conf = len(df_low_conf)

print(f"\n📊 Résumé AfroLID pour les {total_low_conf} cas ambigus (fastText confidence < {CONFIDENCE_THRESHOLD})")
print("-" * 70)


lang_counts_afrolid = df_low_conf["afrolid_lang"].value_counts()
lang_percentages_afrolid = df_low_conf["afrolid_lang"].value_counts(normalize=True) * 100

summary_afrolid_df = pd.DataFrame({
    "Nombre": lang_counts_afrolid,
    "Pourcentage (%)": lang_percentages_afrolid.round(2)
})

print("Distribution des langues détectées par AfroLID :")
display(summary_afrolid_df.head(10))

# Taux de wolof détecté par AfroLID parmi les cas ambigus
# .get() will now work correctly as the index contains strings
wolof_detected_afrolid = lang_counts_afrolid.get("wol", 0)
wolof_rate_afrolid = wolof_detected_afrolid / total_low_conf

print("\n" + "-"*70)
print("📌 Indicateurs clés (AfroLID sur cas ambigus)")
print(f"Instructions détectées comme Wolof par AfroLID : {wolof_detected_afrolid}")
print(f"Taux de Wolof détecté par AfroLID : {wolof_rate_afrolid:.2%}")

# Analyse des scores de confiance d'AfroLID
print("\n" + "-"*70)
print("📈 Statistiques des scores de confiance AfroLID")
display(df_low_conf["afrolid_conf"].describe())

# Impact global estimé
total_global = len(df_wolof)
# We sum up the high confidence wolof detections from fastText (if any) and the low confidence wolof detections from AfroLID
# Assuming df_high_conf was already processed and we know its wolof count
# As per earlier FastText results, wolof_detected for fasttext was 0. So, we only consider wolof_detected_afrolid here.
wolof_global_estimated = wolof_detected_afrolid # Since fastText found 0 wolof at high confidence
wolof_global_rate = wolof_global_estimated / total_global

print("\n" + "-"*70)
print("🌍 Estimation globale après triangulation")
print(f"Wolof détecté (global) : {wolof_global_estimated}")
print(f"Taux global estimé de wolof : {wolof_global_rate:.2%}")

100%|██████████| 2545/2545 [06:34<00:00,  6.46it/s]

✅ Vérification AfroLID des cas ambigus terminée.

📊 Résumé AfroLID pour les 2545 cas ambigus (fastText confidence < 0.7)
----------------------------------------------------------------------
Distribution des langues détectées par AfroLID :


,Nombre,Pourcentage (%)
afrolid_lang,,
wol,1281,50.33
tsn,149,5.85
kin,123,4.83
twi,98,3.85
ibo,89,3.50
swh,69,2.71
zul,59,2.32
fon,59,2.32
lin,57,2.24



----------------------------------------------------------------------
📌 Indicateurs clés (AfroLID sur cas ambigus)
Instructions détectées comme Wolof par AfroLID : 1281
Taux de Wolof détecté par AfroLID : 50.33%

----------------------------------------------------------------------
📈 Statistiques des scores de confiance AfroLID


,afrolid_conf
count,2545.000000
mean,0.001716
std,0.000408
min,0.000176
25%,0.001824
50%,0.001924
75%,0.001928
max,0.001930



----------------------------------------------------------------------
🌍 Estimation globale après triangulation
Wolof détecté (global) : 1281
Taux global estimé de wolof : 28.74%


In [ ]:
df_strong_wolof = df_low_conf[
    (df_low_conf["afrolid_lang"] == "wol") &
    (df_low_conf["afrolid_conf"] >= 0.5)
]

print("Nombre wolof fort :", len(df_strong_wolof))

Nombre wolof fort : 0


In [ ]:
# ===============================
# Distribution complète AfroLID (cas ambigus)
# ===============================

afro_counts_full = df_low_conf["afrolid_lang"].value_counts()
afro_percent_full = df_low_conf["afrolid_lang"].value_counts(normalize=True) * 100

afro_full_summary = pd.DataFrame({
    "Nombre": afro_counts_full,
    "Pourcentage (%)": afro_percent_full.round(2)
})

print("📊 Distribution complète des langues détectées par AfroLID")
display(afro_full_summary)

📊 Distribution complète des langues détectées par AfroLID


,Nombre,Pourcentage (%)
afrolid_lang,,
wol,1281,50.33
tsn,149,5.85
kin,123,4.83
twi,98,3.85
ibo,89,3.50
swh,69,2.71
zul,59,2.32
fon,59,2.32
lin,57,2.24


In [ ]:
# ===============================
# Extraction des textes prédits "wol" par AfroLID
# ===============================

df_afro_wol = df_low_conf[df_low_conf["afrolid_lang"] == "wol"].copy()

print("Nombre total prédits 'wol' par AfroLID :", len(df_afro_wol))

# Colonnes pertinentes pour validation humaine
cols_to_display = [
    "input",
    "ft_pred_lang",
    "ft_confidence",
    "afrolid_lang",
    "afrolid_conf"
]

display(df_afro_wol[cols_to_display].head(20))

Nombre total prédits 'wol' par AfroLID : 1281


,input,ft_pred_lang,ft_confidence,afrolid_lang,afrolid_conf
1,"दुर्घटनापछि, गिब्सनलाई अस्पतालमा लगियो तर केही...",ne,0.626493,wol,0.000714
4,Descrive Cova Figueira. Di ko war a faraaw ci ...,ro,0.151771,wol,0.001921
7,Ci lu tax a ngi leen diw ak liis bu baax ci de...,si,0.139856,wol,0.001927
8,Lu tax lu baax ci ay xaraj yi? Li mu ko diiwal...,eo,0.159266,wol,0.001926
9,"Defeend taariki Senegal. Défaat ci àll bi, ak ...",bn,0.085498,wol,0.001920
14,Ni ko leen di gën a wax? Won la ci niokolo-Kob...,it,0.152316,wol,0.001927
21,"Jaare ci kaw giir yi, te xam-xam na ko gën a b...",en,0.159976,wol,0.001923
22,"Eyi nti, ↄpɛpↄn da a edi kan na wobenya tumi k...",en,0.174856,wol,0.000832
25,Lu tax lu baax ci réew mi? Li mu ngi leen di k...,ia,0.253249,wol,0.001927
30,"Lu tax woon ci réew mi? Li ko gën a waxte, li ...",tr,0.350203,wol,0.001925


In [ ]:
df_afro_wol.to_csv("afrolid_predicted_wolof_low_conf.csv", index=False)

<!-- ===============================
Interprétation — Analyse spécialisée AfroLID
=============================== -->
<h3 style="color:#004080; border-bottom:2px solid #004080; padding-bottom:4px;">
Interprétation des résultats — Analyse AfroLID
</h3>

<p>
L’application d’AfroLID aux 2 545 instances
classées comme ambiguës par fastText
produit une distribution linguistique
plus concentrée sur des langues africaines.
</p>

<p>
Parmi ces instances :
</p>

<ul>
<li>1 281 sont classées comme <strong>wol</strong> (50,33 %)</li>
<li>Les autres langues détectées incluent tsn (5,85 %), kin (4,83 %),
twi (3,85 %), ibo (3,50 %) et plusieurs autres langues africaines
en proportions décroissantes.</li>
</ul>

---

<h4 style="color:#004080;">Analyse des niveaux de confiance</h4>

<p>
Les scores de confiance associés aux prédictions AfroLID
sont extrêmement faibles :
</p>

<ul>
<li>Moyenne : 0,0017</li>
<li>Médiane : 0,0019</li>
<li>Maximum observé : 0,00193</li>
</ul>

<p>
Ces valeurs indiquent que les probabilités attribuées
à la classe prédite sont proches du bruit statistique,
et ne correspondent pas à un signal linguistique fort.
</p>

---

<h4 style="color:#004080;">Interprétation méthodologique</h4>

<p>
Bien que 50,33 % des cas ambigus soient classés comme wolof,
l’extrême faiblesse des scores de confiance
ne permet pas de considérer ces prédictions
comme robustes.
</p>

<p>
L’absence totale de cas détectés comme wolof
avec un niveau de confiance élevé,
y compris après spécialisation,
indique qu’aucun signal linguistique wolof substantiel
n’est observable dans le sous-ensemble étudié.
</p>

<p>
Après triangulation, la proportion globale
d’exemples potentiellement wolof
s’élève à 28,74 % (1 281 / 4 457),
mais sans support probabiliste fort.
</p>

<p>
Ces résultats renforcent l’hypothèse H1
(discussion d’une discordance linguistique)
et fragilisent l’hypothèse H2
(cohérence de validation initiale).
</p>

<!-- ===============================
Analyse comparative et triangulation
=============================== -->
<h2 id="triangulation" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
 Analyse comparative et triangulation
</h2>

<p>
Les résultats obtenus via fastText et AfroLID
présentent des profils contrastés,
nécessitant une analyse comparative structurée.
</p>

---

<h3 style="color:#004080;">Comparaison des signaux détectés</h3>

<ul>
<li>
<strong>fastText (détection globale)</strong> :
0 % d’instances détectées comme wolof,
avec 42,90 % de prédictions à haute confiance
principalement en anglais et en français.
</li>

<li>
<strong>AfroLID (analyse spécialisée des cas ambigus)</strong> :
50,33 % des cas ambigus classés comme wolof,
mais avec des scores de confiance extrêmement faibles
(moyenne ≈ 0,0017).
</li>
</ul>

---

<h3 style="color:#004080;">Évaluation de la robustesse du signal</h3>

<p>
La divergence entre les deux détecteurs
ne réside pas uniquement dans la distribution des classes,
mais dans la qualité probabiliste du signal.
</p>

<p>
Alors que fastText produit des prédictions
à forte probabilité pour une part substantielle du corpus,
AfroLID attribue des probabilités quasi-nulles
à l’ensemble des classes,
y compris pour le wolof.
</p>

<p>
L’absence de détection wolof à haute confiance
dans les deux modèles
indique qu’aucun signal linguistique wolof
robuste et dominant n’est observable.
</p>

---

<h3 style="color:#004080;">Implications pour les hypothèses</h3>

<ul>
<li>
<strong>H1 — Discordance linguistique :</strong>
Les résultats fournissent des éléments empiriques
en faveur d’une discordance significative
entre l’étiquetage « Wolof »
et le contenu effectif du sous-ensemble.
</li>

<li>
<strong>H2 — Cohérence de validation :</strong>
L’absence de confirmation robuste
par des détecteurs indépendants
fragilise l’hypothèse d’une validation initiale cohérente.
</li>

<li>
<strong>H3 — Détectabilité du signal :</strong>
Si un signal wolof majoritaire était présent,
il devrait être détectable avec des probabilités substantielles.
Ce n’est pas le cas dans les résultats observés.
</li>
</ul>

---

<p>
La triangulation des deux approches suggère donc
que le sous-ensemble étudié
ne présente pas de signal wolof détectable
de manière robuste,
malgré son étiquetage initial.
</p>

<!-- ===============================
Discussion méthodologique
=============================== -->
<h2 id="discussion" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Discussion méthodologique
</h2>

<p>
Les résultats obtenus doivent être interprétés
à la lumière des limites inhérentes
aux systèmes de détection automatique de langue.
</p>

---

<h3 style="color:#004080;">Limites liées aux modèles de détection</h3>

<p>
Les classifieurs de langue reposent sur des signaux statistiques
issus de corpus d’entraînement.
Pour les langues à faibles ressources,
la qualité et la représentativité de ces corpus
peuvent influencer la sensibilité du modèle.
</p>

<p>
Toutefois, dans le cas présent,
l’absence de détection wolof à forte probabilité
dans deux détecteurs indépendants
réduit la probabilité qu’un signal wolof majoritaire
soit simplement sous-estimé.
</p>

---

<h3 style="color:#004080;">Effet possible du pipeline Reverse Instruction</h3>

<p>
Le mécanisme de génération par Reverse Instruction
implique plusieurs étapes de traduction
et de transformation textuelle.
Ce processus peut altérer les caractéristiques
linguistiques initiales du document source,
en introduisant des structures syntaxiques
ou lexicales issues de langues pivot,
notamment l’anglais.
</p>

<p>
Une telle transformation pourrait affaiblir
les marqueurs linguistiques distinctifs du wolof,
rendant leur détection plus difficile.
Cependant, même dans ce scénario,
un signal linguistique dominant
devrait rester détectable à un niveau probabiliste substantiel.
</p>

---

<h3 style="color:#004080;">Portée et prudence interprétative</h3>

<p>
Le présent audit ne prétend pas établir
l’absence absolue de contenu en wolof,
mais évalue la robustesse du signal linguistique
au regard de détecteurs indépendants.
</p>

<p>
Les résultats indiquent que,
dans l’état actuel du sous-ensemble analysé,
le signal wolof n’est pas détectable
de manière robuste et dominante.
</p>

<p>
Ces observations invitent à une réévaluation
des procédures de validation linguistique
appliquées lors de la construction du sous-ensemble.
</p>

<!-- ===============================
Conclusion générale
=============================== -->
<h2 id="conclusion" style="color:#004080; border-bottom:3px solid #004080; padding-bottom:5px;">
Conclusion générale
</h2>

<p>
Le présent audit visait à évaluer la cohérence linguistique
du sous-ensemble étiqueté « Wolof »
dans le dataset MURI-IT.
</p>

<p>
L’analyse multilingue globale réalisée avec fastText
montre une domination nette de langues à forte ressource,
notamment l’anglais (43,17 %) et le français (9,96 %),
sans détection du wolof à un niveau de confiance substantiel.
</p>

<p>
L’application d’AfroLID aux cas ambigus
produit une proportion apparente de 50,33 % de classifications « wol »,
mais associée à des probabilités extrêmement faibles
(moyenne ≈ 0,0017),
ne permettant pas de considérer ce signal comme robuste.
</p>

<p>
La triangulation des deux détecteurs indique ainsi
l’absence d’un signal linguistique wolof
dominant et détectable de manière fiable
au sein des 4 457 exemples analysés.
</p>

<p>
Ces résultats apportent un soutien empirique
à l’hypothèse de discordance linguistique (H1),
fragilisent l’hypothèse de cohérence de validation (H2),
et ne confirment pas l’hypothèse de détectabilité robuste du signal (H3).
</p>

<p>
Dans l’état actuel des données,
le sous-ensemble étudié ne présente pas
les caractéristiques attendues
d’un corpus majoritairement rédigé en wolof.
</p>